Proofs/Experiments/AbsVCG.lean

```diff
@@ -96,190 +100,142 @@ theorem Abs.csteps_eq (s : ArmState) (i : Nat) :
-as well as determining the number of steps to simulate in tandem.
--/
-theorem program_effects_lemma (h_pre : abs_pre s0)
-  (h_run : sf = run (Correctness.csteps (stepi s0) 0) (stepi s0)) :
-  r (StateField.GPR 0#5) sf = BitVec.truncate 64
-      (BitVec.zeroExtend 32
-        (BitVec.zeroExtend 64
-          (AddWithCarry (BitVec.zeroExtend 32 (BitVec.zeroExtend 64 (BitVec.zeroExtend 32 (r (StateField.GPR 0#5) s0))))
-              (BitVec.zeroExtend 32
-                (BitVec.truncate 64
-                      (BitVec.replicate 32
-                        (BitVec.extractLsb 31 31 (BitVec.zeroExtend 32 (r (StateField.GPR 0#5) s0)))) &&&
-                    BitVec.truncate 64 (~~~1#32) |||
-                  (BitVec.truncate 64 (BitVec.zero 32) &&& BitVec.truncate 64 (~~~4294967295#32) |||
-                      BitVec.truncate 64 ((BitVec.zeroExtend 32 (r (StateField.GPR 0#5) s0)).rotateRight 31) &&&
-                        BitVec.truncate 64 4294967295#32) &&&
-                    BitVec.truncate 64 1#32))
-              0#1).fst)) ^^^
-    BitVec.truncate 64
-      (BitVec.zeroExtend 32
-        (BitVec.truncate 64
-              (BitVec.replicate 32 (BitVec.extractLsb 31 31 (BitVec.zeroExtend 32 (r (StateField.GPR 0#5) s0)))) &&&
-            BitVec.truncate 64 (~~~1#32) |||
-          (BitVec.truncate 64 (BitVec.zero 32) &&& BitVec.truncate 64 (~~~4294967295#32) |||
-              BitVec.truncate 64 ((BitVec.zeroExtend 32 (r (StateField.GPR 0#5) s0)).rotateRight 31) &&&
-                BitVec.truncate 64 4294967295#32) &&&
-            BitVec.truncate 64 1#32)) ∧
-  r StateField.PC sf = 0x4005e0#64 ∧
-  r StateField.ERR sf = StateError.None := by
-
-  simp only [abs_pre, state_simp_rules] at h_pre
-  have ⟨h_s0_pc, h_s0_program, h_s0_err⟩ := h_pre; clear h_pre
-
-  -- Instruction 1
-  rw [Abs.csteps_eq, abs_stepi_0x4005d0_cut s0 h_s0_program h_s0_pc h_s0_err] at h_run
-  simp only [minimal_theory, Nat.reduceAdd] at h_run
-  generalize h_step_1 : stepi s0 = s1 at h_run
-  have h_s1 : s1 = run 1 s0 := by
-    simp only [run, h_step_1]
-  replace h_step_1 : s1 = stepi s0 := h_step_1.symm
-  rw [program.stepi_0x4005d0 s0 s1 h_s0_program h_s0_pc h_s0_err] at h_step_1
-  have h_s1_program : s1.program = program := by
-    simp only [h_step_1, h_s0_program,
-               state_simp_rules, minimal_theory, bitvec_rules]
-  have h_s1_err : r StateField.ERR s1 = StateError.None := by
-    simp only [h_step_1, h_s0_err,
-               state_simp_rules, minimal_theory, bitvec_rules]
-  have h_s1_pc : r StateField.PC s1 = 0x4005d4#64 := by
-    simp only [h_step_1, h_s0_pc,
-               state_simp_rules, minimal_theory, bitvec_rules]
+theorem nextc_to_run_from_0x4005d0 (h : abs_pre s0) :
+  (Correctness.nextc (Sys.run s0 1)) = run 4 s0 := by
+  simp only [abs_pre, state_simp_rules] at h
+  obtain ⟨h_s0_pc, h_s0_program, h_s0_err, h_s0_sp_aligned⟩ := h
+  simp only
+    [Correctness.nextc, Correctness.arm_run,
+     Spec'.cut, minimal_theory]
 
```

Comment:
Yes, that's exactly right. And I'm glad this appears repetitive --- that was the goal, to see if there's a pattern that can be mechanized.

I'm hoping we can avoid such `nextc_to_run..` theorems entirely, and provide automation for theorems like `effects_of_nextc_..` directly where the number of steps to simulate and the effects are determined "dynamically" in tandem. It'd be useful to build support for this *atop* the existing symbolic eval machinery so that we're not married to the VCG approach for all LNSym proofs. E.g., the VCG approach is pretty tedious for verifying straight-line code but convenient for loops. For the former, we could use the existing symbolic eval machinery to simulate the code and then immediately reason about the program's effects. For the latter, we could use the VCG-supporting symbolic eval machinery.